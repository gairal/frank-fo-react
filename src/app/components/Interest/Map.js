import React, { Component } from 'react';
import PropTypes from 'prop-types';
import GoogleMap from 'react-google-maps/lib/components/GoogleMap';
import Marker from 'react-google-maps/lib/components/Marker';
import { withScriptjs, withGoogleMap } from 'react-google-maps/lib';

const GMap = withScriptjs(
  withGoogleMap(({ children, center, zoom }) => (
    <GoogleMap zoom={zoom} center={center}>
      {children}
    </GoogleMap>
  )),
);

export default class Map extends Component {
  static propTypes = {
    travels: PropTypes.arrayOf(PropTypes.shape()).isRequired,
  };

  constructor(props) {
    super(props);

    this.state = {
      markers: [],
      center: { lat: 48.856149, lng: 2.3386383 },
      zoom: 8,
    };

    this.isAlive = true;
  }

  componentDidUpdate() {
    const { travels } = this.props;
    const { markers } = this.state;
    if (travels.length && !markers.length) {
      this.travel(0);
    }
  }

  componentWillUnmount() {
    this.isAlive = false;
  }

  travel(index) {
    const { travels } = this.props;
    if (this.isAlive) {
      const currTravel = travels[index];
      if (!currTravel) {
        this.setState(prevState => ({
          ...prevState,
          zoom: 2,
        }));

        return;
      }

      const travel = {
        lat: currTravel.coordinates.Lat,
        lng: currTravel.coordinates.Lng,
      };

      this.setState(prevState => ({
        ...prevState,
        markers: prevState.markers.concat(travel),
        center: travel,
      }));

      setTimeout(() => this.travel(index + 1), 2000);
    }
  }

  render() {
    const { markers, center, zoom } = this.state;
    return (
      <GMap
        googleMapURL="https://maps.googleapis.com/maps/api/js?key=AIzaSyALPF75dwSUZIWeAM-eBvMNA4BN3p-cCj4"
        loadingElement={<div style={{ height: '100%' }} />}
        containerElement={<div style={{ height: '400px' }} />}
        mapElement={<div style={{ height: '100%' }} />}
        center={center}
        zoom={zoom}
      >
        {markers.map(e => (
          <Marker key={e.lat} position={e} />
        ))}
      </GMap>
    );
  }
}
