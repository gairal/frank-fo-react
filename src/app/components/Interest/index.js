import interest from './Interest';

export default interest;
