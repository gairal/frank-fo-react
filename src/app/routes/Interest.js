import component from '@/components/Interest';
import AbstractRoute from './AbstractRoute';

export default class Interest extends AbstractRoute {
  constructor(store) {
    super(store, 'interests', 'language', 'categories');
    this.component = component;
    this.initialState.travels = [];

    this.ACTIONS = {
      FETCH_SUCCESS: 'INTEREST_FETCH_SUCCESS',
      FETCH_FAILURE: 'INTEREST_FETCH_FAILURE',
    };

    this.init();
  }

  // actions
  static success({ json, travels }) {
    return {
      type: 'INTEREST_FETCH_SUCCESS',
      payload: {
        json,
        travels,
      },
    };
  }

  static fail(err) {
    return {
      type: 'INTEREST_FETCH_FAILURE',
      payload: {
        err,
      },
    };
  }

  static fetchSuccess(state, action) {
    return ({
      ...state,
      data: action.payload.json,
      travels: action.payload.travels,
    });
  }

  loadFactory() {
    const API_URLS = [
      this.API_URL,
      'https://com-gairal-frank-bo.appspot.com/travels',
    ];

    return () => (dispatch) => {
      dispatch(AbstractRoute.showLoader());

      const fetchOptions = {
        method: 'GET',
      };

      return Promise.all(API_URLS.map(url => fetch(url, fetchOptions)))
        .then(responses =>
          Promise.all(responses.map(res => res.json()))
            .then((jsons) => {
              const payload = jsons.reduce((acc, e) => {
                if (e[0].name) {
                  return { ...acc, json: e };
                }
                return { ...acc, travels: e };
              }, {});

              dispatch(Interest.success(payload));
              return dispatch(AbstractRoute.hideLoader());
            }))
        .catch((err) => {
          dispatch(Interest.fail(err));
          return dispatch(AbstractRoute.hideLoader());
        });
    };
  }

  init() {
    super.init();

    this.ACTION_HANDLERS[this.ACTIONS.FETCH_SUCCESS] = Interest.fetchSuccess;
    this.mapStateToProps.travels = 'travels';
  }
}
