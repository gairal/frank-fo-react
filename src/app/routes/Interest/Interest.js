import AbstractRoute from '../AbstractRoute';
import component from '../../components/Interest';

export default class Interest extends AbstractRoute {
  constructor(store) {
    super(store, 'interests', 'categories');
    this.component = component;
    this.initialState.travels = [];

    this.ACTION_HANDLERS[this.ACTIONS.FETCH_SUCCESS] = (state, action) => ({
      ...state,
      data: action.payload.json,
      travels: action.payload.travels,
      isFetching: false,
    });

    this.mapStateToProps.travels = 'travels';
  }

  static success({ json, travels }) {
    return {
      type: 'FETCH_SUCCESS',
      payload: {
        json,
        travels,
      },
    };
  }

  loadFactory() {
    const API_URLS = [
      this.API_URL,
      'https://api-dot-com-gairal-frank.appspot.com/travels',
    ];

    return () => (dispatch) => {
      dispatch(AbstractRoute.request());

      const fetchOptions = {
        method: 'GET',
      };

      return Promise.all(API_URLS.map(url => fetch(url, fetchOptions)))
        .then(responses =>
          Promise.all(responses.map(res => res.json()))
            .then((jsons) => {
              const payload = jsons.reduce((acc, e) => {
                if (e[0].name) {
                  return { ...acc, json: e };
                }
                return { ...acc, travels: e };
              }, {});

              return dispatch(Interest.success(payload));
            }));
    };
  }
}
