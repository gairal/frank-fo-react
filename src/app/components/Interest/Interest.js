import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {
  Divider,
  List,
  ListItem,
  ListItemText,
  ListSubheader,
} from '@material-ui/core';
import Map from './Map';

export default class Interest extends Component {
  static propTypes = {
    load: PropTypes.func.isRequired,
    interests: PropTypes.arrayOf(PropTypes.shape()),
    travels: PropTypes.arrayOf(PropTypes.shape()),
  };

  static defaultProps = {
    interests: [],
    travels: [],
  };

  componentDidMount() {
    const { load } = this.props;
    load();
  }

  render() {
    const { travels, interests } = this.props;

    const listItem = interest => (
      <ListItem key={interest.name}>
        <ListItemText primary={interest.name} />
      </ListItem>
    );

    return (
      <section>
        <Map travels={travels} />
        <List>
          {interests.map(category => (
            <li key={category.name}>
              <ul>
                <ListSubheader>{category.name}</ListSubheader>
                {category.interests ? category.interests.map(listItem) : null}
              </ul>
              <Divider />
            </li>
          ))}
        </List>
      </section>
    );
  }
}
