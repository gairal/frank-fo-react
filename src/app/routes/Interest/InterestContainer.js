import { connect } from 'react-redux';
import { add, remove } from '../Interest';

import Experience from '../../components/Interest';

const mapDispatchToProps = {
  add,
  remove,
};

const mapStateToProps = state => ({
  todos: state.todo,
});

export default connect(mapStateToProps, mapDispatchToProps)(Experience);
