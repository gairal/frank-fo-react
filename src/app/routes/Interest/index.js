import Interest from './Interest';

export default store => new Interest(store).route;
