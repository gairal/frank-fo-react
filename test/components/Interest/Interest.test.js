import React from 'react';
import renderer from 'react-test-renderer';
import Interest from '@/components/Interest';

describe('components > Interest', () => {
  it('Renders correctly default', () => {
    const component = renderer.create(<Interest
      load={jest.fn}
      interests={[{ name: 'test' }]}
      travels={[
        {
          place: 'test',
        },
      ]}
    />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Renders correctly with interests', () => {
    const component = renderer.create(<Interest
      load={jest.fn}
      interests={[
        {
          name: 'test',
          interests: [
            {
              name: 'test',
              title: 'test',
              place: 'test',
              date_in: '2018-01-01',
              description: 'test',
              image: { name: 'test' },
            },
          ],
        },
      ]}
      travels={[
        {
          place: 'test',
        },
      ]}
    />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
