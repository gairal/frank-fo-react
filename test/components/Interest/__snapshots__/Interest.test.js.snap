// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components > Interest Renders correctly default 1`] = `
<section>
  <div
    style={
      Object {
        "height": "100%",
      }
    }
  />
  <ul
    className="MuiList-root-1 MuiList-padding-2"
  >
    <li>
      <ul>
        <li
          className="MuiListSubheader-root-5 MuiListSubheader-sticky-10 MuiListSubheader-gutters-8"
        >
          test
        </li>
      </ul>
      <hr
        className="MuiDivider-root-11"
      />
    </li>
  </ul>
</section>
`;

exports[`components > Interest Renders correctly with interests 1`] = `
<section>
  <div
    style={
      Object {
        "height": "100%",
      }
    }
  />
  <ul
    className="MuiList-root-1 MuiList-padding-2"
  >
    <li>
      <ul>
        <li
          className="MuiListSubheader-root-5 MuiListSubheader-sticky-10 MuiListSubheader-gutters-8"
        >
          test
        </li>
        <li
          className="MuiListItem-root-15 MuiListItem-default-18 MuiListItem-gutters-22"
          disabled={false}
        >
          <div
            className="MuiListItemText-root-25"
          >
            <span
              className="MuiTypography-root-31 MuiTypography-subheading-38 MuiListItemText-primary-28"
            >
              test
            </span>
          </div>
        </li>
      </ul>
      <hr
        className="MuiDivider-root-11"
      />
    </li>
  </ul>
</section>
`;
